pydarn/pydmap/io.py

```diff
@@ -0,0 +1,1296 @@
+                message = "Error: Array {index}-dimension size {size}"\
+                        " exceeds record size: {rec_size}."\
+                        "".format(index=i,
+                                  size=array_shape[i],
+                                  rec_size=record_size)
+                raise pydmap_exceptions.DmapDataError(self.dmap_file, message)
+
+        # We could use np.prod(array_shape) but the for loop has a better
+        # time performance.
+        total_num_cells = 1
+        for i in array_shape:
+            total_num_cells *= i
+        self.bytes_check(total_num_cells, "total number of cells",
+                         record_size, "record size")
+
+        total_num_cells_bytes = total_num_cells * array_fmt_bytes
+        self.bytes_check(total_num_cells_bytes,
+                         "total number of cells in bytes",
+                         record_size, "record size")
+
+        # parsing an array of strings requires a different method. Numpy can't
+        # parse strings or dmaps into arrays the way it can for other
+        # types because it doesn't
+        # know the sizes. They have to be manually read the slow way.
+        # Because chars
+        # are encoded as hex literals, they have to be read one at a
+        # time to make sense.
+
+        # what is DMAP format? Can we get an example that has a
+        # list of DMAPs?
+        # Also why is char also included?
+        if array_type_fmt == 's' or array_type_fmt == 'c':
+            array_value = self.read_string_array(array_shape,
+                                                 array_type_fmt,
+                                                 array_fmt_bytes)
+        else:
+            array_value = self.read_numerical_array(array_type_fmt,
+                                                    total_num_cells,
+                                                    array_fmt_bytes)
+
+        return DmapArray(array_name, array_value, array_type,
+                         array_dimension, array_shape)
+
+    def read_data(self, data_type_fmt, data_fmt_bytes: int):
+        """
+        Reads in individual data type from the byte array
+        Given a dmap data type.
+
+        Parameter
+        ---------
+        data_type : int
+            dmap data type numerical value for knowing how many bytes to read.
+
+        Return
+        ------
+        data : data_type
+            returns the data that was store in the byte array.
+            Has the data type of the dmap data type passed in.
+
+        Raises
+        ------
+        CursorError
+            if cursor is not correctly aligned or at the correct position.
+        DmapDataError
+            if trying to read a DMAP data type
+
+        See Also
+        --------
+        Struct.unpack : converts the bytes array values from a C structure to
+        python objects
+
+        See DEVELOPER_README.md for more information on DMAP data types and
+        Struct.unpack.
```

Comment:
I'm not sure I ever tested numpy casting vs struct.unpack here. Perhaps some performance benefits in using numpy.

Reply:
Nope, I tested them. 

Numpy only does well at reading in numerical arrays. If you get it to read in a single value or an array of strings similar to how you do it with struct.unpack it is slower and then on top of that numpy.frombuffer method returns it in an array so you need to call another method to get the string out or the single value which adds to the overhead :(